### PR TITLE
add freebsd support, prepare for other BSDs

### PR DESCRIPTION
**Pull request by wucke13**

### CODE REVIEW

**Comment by emilengler** [comment]
```
engler@thecure rosenpass % shellcheck rp

In rp line 213:
  case `uname` in
       ^-----^ SC2006 (style): Use $(...) notation instead of legacy backticks `...`.

Did you mean:
  case $(uname) in


In rp line 240:
    *)
    ^-- SC2221 (warning): This pattern always overrides a later one on line 244.


In rp line 241:
      fatal "Your system `uname` is not yet supported. We are happy to receive patches to address this :)"
                         ^-----^ SC2006 (style): Use $(...) notation instead of legacy backticks `...`.

Did you mean:
      fatal "Your system $(uname) is not yet supported. We are happy to receive patches to address this :)"


In rp line 244:
    NetBSD)
    ^----^ SC2222 (warning): This pattern never matches because of a previous pattern on line 240.

For more information:
  https://www.shellcheck.net/wiki/SC2221 -- This pattern always overrides a l...
  https://www.shellcheck.net/wiki/SC2222 -- This pattern never matches becaus...
  https://www.shellcheck.net/wiki/SC2006 -- Use $(...) notation instead of le...
```

**Comment by emilengler** [comment]
@wucke13 Forgive me the force-push, but I have made these fixes myself branch.

File: rp [hunk 1]
```diff
@@ -209,15 +209,47 @@ exchange() {
     fatal "Needs at least one peer specified"
   fi
 
-  frag "
-    # Create the Wireguard interface
-    ip link add dev $(enquote "${dev}") type wireguard || true"
+  # os dependent setup
+  case $(uname) in
```
Comment (koraa) [inline rp:213]: Nitpick: Could we move that into a function?

`setup_device_frag()`

Comment (koraa) [inline rp:213]: How about `"$(uname)"`?

File: rp [hunk 2]
```diff
@@ -209,15 +209,47 @@ exchange() {
...
+        ip link del dev $(enquote "${dev}") || true"
 
-  frag "
-    ip link set dev $(enquote "${dev}") up"
+      frag "
+        ip link set dev $(enquote "${dev}") up"
+      ;;
+
+    FreeBSD)
+      kldload -n if_wg || fatal "Cannot load if_wg kernel module"
```
Comment (koraa) [inline rp:227]: Can we move the `kldload` into the frag?

**Review by koraa — APPROVED** [review]
I kind of would like the `kdload` and the `"$(uname)"` to be addressed before merge.

Other than that this looks great!

File: rp (hunk 1 above)
[the remark this answers is not in the document]
Comment (wucke13) [reply rp:213]: Didn't know about it, will use it. Though it creates some caveats, i.e. the stackoverflow you linked mentioned only `linux-gnu`, which for sure was upsetting the alpine user next to me.

**Comment by wucke13** [comment]
@koraa @emilengler Thank you for the constructive feedback, I tried to incorporate it accordingly.

**Comment by emilengler** [comment]
Can we merge this and make a release ASAP?

**Comment by wucke13** [comment]
# Todo

- [ ] get rid of the PATH replacement for finding the rosenpass binary
   - [ ] Or at least get rid of GNU grep extensions and git warnings such as `WARNING: Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).`
- [ ] IPv6 Handing of colon in ip for `endpoint` verb in `rp` CLI
- [ ]

**Comment by emilengler** [comment]
```diff
diff --git a/rp b/rp
index a9d674e..b668de7 100755
--- a/rp
+++ b/rp
@@ -223,7 +223,7 @@ exchange() {
         ip link set dev $(enquote "${dev}") up"
       ;;

-    freebsd)
+    freebsd*)
       frag "
         # load the WireGuard kernel module
         kldload -n if_wg || fatal 'Cannot load if_wg kernel module'"
@@ -352,7 +352,9 @@ main() {
   verbose=0
   scriptdir="$(dirname "${script}")"
   gitdir="$(detect_git_dir)" || true
-  nixdir="$(readlink -f result/bin/rp | grep -Pio '^/nix/store/[^/]+(?=/bin/[^/]+)')" || true
+  if [[ -d /nix ]] then
+    nixdir="$(readlink -f result/bin/rp | grep -Pio '^/nix/store/[^/]+(?=/bin/[^/]+)')" || true
+  fi
   binary="$(find_rosenpass_binary)"

   # Parse command
```

**Review by emilengler — APPROVED** [review]

**Comment by wucke13** [comment]
The shell script might be working, but we see occasional kernel crashes in our FreeBSD Aarch64 VM.

**Pull request merged**